proxy.go

```diff
@@ -289,7 +289,18 @@ func (rp *reverseProxy) serveFromCache(s *scope, srw *statResponseWriter, req *h
 			cacheMissFromConcurrentQueries.With(labels).Inc()
 			log.Debugf("%s: cache miss after awaiting concurrent query", s)
 		}
-
+	} else {
+		err = fmt.Errorf("%s: no result found during grace time period : %s", s)
+		// Request has awaited longer than half of its timeout.
+		// There are following options:
+		// - the identical query failed to register its completion (TransactionRegistry unavailable)
+		// - configured grace time was too short
+		//
+		// Because of those 2 options, we will fail the query if elapsed time is bigger than half of the max request timeout.
+		if transactionResult.ElapsedTime > s.user.maxExecutionTime / 2 {
+			respondWith(srw, err, http.StatusRequestTimeout)
```

Comment:
don't you think it may be weird for the end user to receive timeouts after 15 seconds ou 30 seconds ?

Reply:
You're right. Perhaps it should be treated more as an internal server error. What do you think?

Reply:
yes and a message like "Won't be able to server your query within ..." ?

Reply:
becareful because on new clickhouse versions (including the one we have in production), clickhouse will timeout after a few seconds if the forecasted execution time of the query is above the max_execution_time.

Reply:
Very good point. I'd propose that we rely on the idea of state machine in `transactionRegistry`. We would check the state for the query (e.g. Created, Failed, Completed) and take a reliable decision based on that. What do you think?

Reply:
looks good, as long as the state contains a Timeout state (because the Failed state could be due to other factors)